detect encryption of ed25519 ssh keys

ostrowr opened the pull request:
Newer versions of OpenSSH use ed25519 keys, which don't include the text "ENCRYPTED" in their body. This patch ~checks to see if the first few bytes of the key line up with what we would expect from an unencrypted key.~ tries to parse the keys using openssl to support arbitrary key types.

Fixes #4826 
Fixes #5253. 
Fixes #4886

~I haven't added tests as I haven't seen any for the SSH table yet, but I'm happy to look into that if helpful.~ Also added a bunch of tests for the user_ssh_keys table.

directionless commented:
ed25519 are becoming more common. Any chance at having this reviewed?

guliashvili commented on osquery/tables/system/posix/ssh_keys.cpp:
```diff
@@ -16,10 +16,31 @@
+  if (boost::starts_with(keys_content, kEd25519Header)) {
+    const std::string prefix = keys_content.substr(
+        kEd25519Header.size() + 1, kEd25519UnencryptedPrefix.size());
+    return prefix != kEd25519UnencryptedPrefix;
```
auto keys_body = keys_content.substr(kEd25519Header.size() + 1)
return boost::starts_with(keys_body, kEd25519UnencryptedPrefix)

guliashvili replied:
Thanks for resolving the issue.

I'm afraid substr(
        kEd25519Header.size() + 1, kEd25519UnencryptedPrefix.size()) can crash the code if string is smaller. Does my suggestion look okay?

ostrowr replied:
Whoops, it's been a long time since I've written any C-esque code.

I'll push up a fix momentarily.

ostrowr replied:
Ah, actually `.substr` does do what I thought – it just gets characters to the end of the string if the second argument is greater than the length. However, this will still crash if the string is exactly `kEd25519Header` since substr throws if the first argument is greater. So your suggestion made me notice a bug even though it doesn't quite work!

ostrowr commented:
Changed to have the header include a newline, so the substr call is guaranteed to succeed. Also updated the isEncrypted function to take a string reference as that seems to be the pattern throughout the codebase.

ostrowr commented:
Let me know if there's anything else I can do here!

faf0-addepar commented on osquery/tables/system/posix/ssh_keys.cpp:
```diff
@@ -16,10 +16,31 @@
+        kEd25519Header.size(), kEd25519UnencryptedPrefix.size());
+    return prefix != kEd25519UnencryptedPrefix;
+  }
+  return (keys_content.find("ENCRYPTED") != std::string::npos) ? true : false;
```
```suggestion
  return keys_content.find("ENCRYPTED") != std::string::npos;
```

is that sufficient?

ostrowr replied:
:+1:

szippy commented:
Any update on when this can be merged?

theopolis commented:
Will take a look shortly, thanks for the update!

directionless commented:
I'd love to see tests here.

ostrowr commented:
Haven't forgotten about this; hoping to have a few minutes to write some tests this weekend!

theopolis commented:
Hi @ostrowr, any updates on this?

ostrowr commented:
Sorry for the lengthy delay – traveling and working! Was planning on waiting until 4.0.1 so the new build system was well-defined, but if this is something you'd like to get in before the end of the month, I'm happy to take care of it this weekend (really, this time.)

ostrowr commented on osquery/tables/system/BUCK:
```diff
@@ -32,6 +32,7 @@ osquery_cxx_library(
             [
                 "posix/known_hosts.h",
                 "posix/shell_history.h",
+                "posix/ssh_keys.h",
```
not sure if I need to modify buck files elsewhere; was struggling to get the buck build working.

ostrowr commented on osquery/tables/system/tests/posix/ssh_keys_tests.cpp:
```diff
@@ -0,0 +1,283 @@
+  EXPECT_EQ(row.at("key_type"), "");
+}
+
+TEST_F(SshKeysTests, ed25519_unencrypted) {
```
These ed25519 tests are currently failing because the version of openssl that we bundle doesn't parse them properly. We can either update the openssl version or fall back to the brittle ssh key parsing code from the previous revision.

theopolis replied:
Could we keep the current implementation in the PR but check https://www.openssl.org/docs/manmaster/man3/OPENSSL_VERSION_NUMBER.html within the tests and only run this test for the correct version?

ostrowr replied:
Turns out the new version doesn't work either! Now we use openssl if possible, but if it can't read the key we fall back to naive parsing.

ostrowr commented on osquery/tables/system/posix/ssh_keys.h:
```diff
@@ -0,0 +1,32 @@
+/**
```
Should this header still be Facebook?

theopolis replied:
We'll change it later on in a codemod.

ostrowr commented:
I've updated this table to use the openssl .pem parsing, which was already bundled in osquery so there's no extra cost to doing so.

Also added a column "key_type" which is reported if openssl can grok it.

This should make parsing SSH keys a lot less brittle, but still doesn't quite support ed25519 keys. I'm 90% sure that an upgrade of openssl versions will work, but if this isn't possible, we can revert to https://github.com/osquery/osquery/commit/89f64baca57f3c458ddf1e3912ccb316943fb25b but keep the tests.

ostrowr commented:
Happy to go either direction here; whatever the maintainers are more comfortable with!

theopolis commented:
It might take some time to update the OpenSSL version but we can make that happen. We are on the path to migrating dependencies to not be pre-built.

It sounds like you will be blocked until that occurs?

ostrowr commented:
I can check against a hardcoded magic value until then (see the messages above) so we can at least get this feature working, and make it a little more robust in a future PR.

I haven’t been following the dependency updates too much but happy to help there as well.

On Fri, Sep 20 2019 at 9:31 PM, < notifications@github.com > wrote:

> 
> 
> 
> It might take some time to update the OpenSSL version but we can make that
> happen. We are on the path to migrating dependencies to not be pre-built.
> 
> 
> 
> It sounds like you will be blocked until that occurs?
> 
> 
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub (
> https://github.com/osquery/osquery/pull/5526?email_source=notifications&email_token=AAYWP3IUL5YWIMCHUZXL3LLQKWPSLA5CNFSM4G6JNPM2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7IKEKA#issuecomment-533766696
> ) , or mute the thread (
> https://github.com/notifications/unsubscribe-auth/AAYWP3OWOHIRGHBG7CINLYDQKWPSLANCNFSM4G6JNPMQ
> ).
> 
> 
>

theopolis commented on osquery/tables/system/posix/ssh_keys.h:
```diff
@@ -0,0 +1,27 @@
+namespace osquery {
+namespace tables {
+
+const std::string kSSHUserKeysDir = ".ssh/";
```
Can you move the definition into an implementation file?

theopolis commented on osquery/tables/system/posix/ssh_keys.cpp:
```diff
@@ -13,13 +13,84 @@
+
+  EVP_PKEY* pkey;
+  pkey =
+      PEM_read_bio_PrivateKey(bio_stream, nullptr, passwordCallback, nullptr);
```
Can you write `auto pkey = ` instead of declaring `pkey` on the above line?

theopolis commented:
If this is working for CMake I can help with the Buck changes. 

I think we should keep the current version (vs. falling back to the brittle version) and try to update OpenSSL within the dependencies at the same time.

ostrowr commented:
After only a year, this is ready for re-review!

Smjert commented:
@ostrowr Hi! Thanks for your work and sorry for the delay.
This needs to be rebased on latest master because we've dropped Buck and we also now have openssl 1.1.1g, so something may have changed.

theopolis commented:
This data is valuable. I may have time soon to help out if needed. Let me know.

ostrowr commented:
I'm hoping to have a chance to get to this in the next week or so.

theopolis commented:
Ping, how is this coming along, do you need any help?

rhuddleston commented:
Any update on this PR? Would be really nice to get this fixed up
